src/S3/S3Transfer/S3TransferManager.php

```diff
@@ -0,0 +1,850 @@
+            );
+        }
+
+        if ($this->requiresMultipartUpload($uploadRequest->getSource(), $mupThreshold)) {
+            return $this->tryMultipartUpload(
+                $uploadRequest,
+                $listenerNotifier
+            );
+        }
+
+        return $this->trySingleUpload(
+            $uploadRequest->getSource(),
+            $uploadRequest->getUploadRequestArgs(),
+            $listenerNotifier
+        );
+    }
+
+    /**
+     * @param UploadDirectoryRequest $uploadDirectoryRequest
+     *
+     * @return PromiseInterface
+     */
+    public function uploadDirectory(
+        UploadDirectoryRequest $uploadDirectoryRequest,
+    ): PromiseInterface
+    {
+        $uploadDirectoryRequest->validateSourceDirectory();
+        $targetBucket = $uploadDirectoryRequest->getTargetBucket();
+
+        $uploadDirectoryRequest->updateConfigWithDefaults(
+            $this->config->toArray()
+        );
+
+        $config = $uploadDirectoryRequest->getConfig();
+        $progressTracker = $uploadDirectoryRequest->getProgressTracker();
+        if ($progressTracker === null
+            && ($config['track_progress'] ?? $this->config->isTrackProgress())) {
+            $progressTracker = new MultiProgressTracker();
+        }
+
+        $filter = null;
+        if (isset($config['filter'])) {
+            if (!is_callable($config['filter'])) {
+                throw new InvalidArgumentException(
+                    "The provided config `filter` must be callable."
+                );
+            }
+
+            $filter = $config['filter'];
+        }
+
+        $uploadObjectRequestModifier = null;
+        if (isset($config['upload_object_request_modifier'])) {
+            if (!is_callable($config['upload_object_request_modifier'])) {
+                throw new InvalidArgumentException(
+                    "The provided config `upload_object_request_modifier` must be callable."
+                );
+            }
+
+            $uploadObjectRequestModifier = $config['upload_object_request_modifier'];
+        }
+
+        $failurePolicyCallback = null;
+        if (isset($config['failure_policy'])) {
+            if (!is_callable($config['failure_policy'])) {
+                throw new InvalidArgumentException(
+                    "The provided config `failure_policy` must be callable."
+                );
+            }
+
+            $failurePolicyCallback = $config['failure_policy'];
+        }
+
+        $sourceDirectory = $uploadDirectoryRequest->getSourceDirectory();
+        $dirIterator = new RecursiveDirectoryIterator(
+            $sourceDirectory
+        );
+
+        $flags = FilesystemIterator::SKIP_DOTS;
+        if ($config['follow_symbolic_links'] ?? false) {
+            $flags |= FilesystemIterator::FOLLOW_SYMLINKS;
+        }
+
+        $dirIterator->setFlags($flags);
+
+        if ($config['recursive'] ?? false) {
+            $dirIterator = new RecursiveIteratorIterator(
```

Comment:
I would name `RecursiveDirectoryIterator` and `RecursiveIteratorIterator` separate things

Reply:
I don`t think is necessary to create two variables for this. At the end we just need to iterate over a single iterator that can be either a `RecursiveDirectoryIterator` or a `RecursiveIteratorIterator`. So having a single variable that will be one of those based on those conditions there I think should be fine.